Fixed various bugs in Table and TableView aggregate methods

rrrlasse opened the pull request:
Fixes https://github.com/realm/realm-core/issues/2357 and other aggregate issues

rrrlasse commented:
@realm/core please review, I heard the fix should be urgent

beeender commented:
@cmelchior I am not sure if your issue happens with same conditions, maybe you can have a check to see if they are the same?

beeender commented:
And i might meet some similar issue on other aggregate cases, but i thought it was my fault at that time. Could it be possible the similar issue happen to others? I saw some thing similar with `max_int` earlier yesterday.

finnschiermer commented on src/realm/timestamp.hpp:
```diff
@@ -97,22 +97,24 @@ class Timestamp {
-        REALM_ASSERT(!is_null());
-        REALM_ASSERT(!rhs.is_null());
-        return m_seconds != rhs.m_seconds || m_nanoseconds != rhs.m_nanoseconds;
+        return is_null() != rhs.is_null() || m_seconds != rhs.m_seconds || m_nanoseconds != rhs.m_nanoseconds;
```
This is tricky. Much cleaner to implement != from == :-)

rrrlasse replied:
oh, very true, will fix that :)

ironage commented:
@rrrlasse Please consider the following two cases. I haven't tested these myself but from the code it seems like the first element will be returned and I think it should be not found (`npos`). Either way can you add test cases for them?
1) `minmax` on a nullable timestamp column of size = 1 with the first element being null
2) `minmax` on a nullable timestamp column of size > 1 with all elements being null

beeender commented:
This is not only happened to timestamp but also other column types, for example:

```cpp
ONLY(TableView_MaxBug)
{
    Table table;
    table.add_column(type_Int, "value", true);
    table.add_empty_row(4);
    table.set_int(0, 0, 1);
    table.set_int(0, 1, 2);
    table.set_null(0, 2);
    table.set_int(0, 3, 42);

    auto tv = table.where().not_equal(0, 42).find_all();
    CHECK_EQUAL(tv.size(), 3);
    CHECK_EQUAL(tv.maximum_int(0), 2);
}
```

This will crash with 
> test_table_view.cpp:3434: ERROR in TableView_MaxBug: Unhandled exception realm::util::BadOptionalAccess: bad optional access

rrrlasse commented:
I have fixed numerous issues in aggregate methods on Table and TableView. I'm not sure I've catched them all but wanted to provide a quick fix since it was an urgent issue. Will look deeper into it on monday.

All this code must be cleaned up as it's super brittle and messy. Problem is that we have the same methods on both Table, TableView and Query, and all versions have all kinds of speed-optimization pathways for special cases (such as executing method on Column if Query had no search condition, etc, etc, etc.).

We need to simplify all that.

rrrlasse commented:
@ironage I have added TEST(Table_MinMaxSingleNullRow) that documents behaviour for that case.

The return value is undefined for columns with only null-values. You must always check the return_ndx parameter instead. It's an old leftover from before we introduced nulls in Realm. Should be fixed one day....

danielpovlsen commented on src/realm/column_timestamp.hpp:
```diff
@@ -126,20 +126,21 @@ class TimestampColumn : public ColumnBaseSimple {
 
         Timestamp best = get(0);
-        size_t best_index = 0;
+        size_t best_index = !is_null(0) ? 0 : npos; // npos if null
```
Suggestion:
```C++
size_t best_index = best.is_null() ? npos : 0;
```

rrrlasse replied:
Not possible when `best` is of a primitive C++ type

rrrlasse replied:
True, should be faster. Disregard my deleted message

rrrlasse commented:
@realm/core @beeender please review

ironage commented on CHANGELOG.md:
```diff
@@ -1,5 +1,15 @@
 # 2.2.1 Release notes
 
+### Bugfixes
```
This will merge to the wrong section now that we have released again. Don't forget to update to the latest master and merge these notes to the next release section.

danielpovlsen replied:
@rrrlasse Was this addressed? (cc: @finnschiermer)

danielpovlsen replied:
It looks like @finnschiermer fixed this in #2373, but please address all review feedback before merging.

ironage commented on src/realm/table_view.cpp:
```diff
@@ -356,9 +375,9 @@ double TableViewBase::minimum_double(size_t column_ndx, size_t* return_ndx) cons
+        return aggregate<act_Min, int64_t>(&IntNullColumn::minimum, column_ndx, 0, return_ndx);
     else
-        return aggregate<act_Max, int64_t>(&IntegerColumn::minimum, column_ndx, 0, return_ndx);
+        return aggregate<act_Min, int64_t>(&IntegerColumn::minimum, column_ndx, 0, return_ndx);
```
Wow, good eye! I'm a bit surprised our tests didn't catch it.

rrrlasse replied:
If the TableView size matched the Table size, the bug got bypassed by executing the correct "min" method directly on the Column. This fix is irrelevant today though because this old datetime code is unreachable except through unit test

ironage commented on src/realm/table_view.cpp:
```diff
@@ -185,84 +191,96 @@ R TableViewBase::aggregate(R (ColType::*aggregateMethod)(size_t, size_t, size_t,
     const ColType* column = static_cast<ColType*>(&m_table->get_column_base(column_ndx));
 
+    // FIXME: Optimization temporarely removed for stability
+/*
```
I would prefer that you remove all this optimisation code instead of commenting it out, because it could take some larger redesign to optimise again and it will most likely be in a different way. But it's not blocking for me.

rrrlasse commented:
@finnschiermer approve review?

bmunkholm commented on CHANGELOG.md:
```diff
@@ -2,8 +2,10 @@
 
-* Lorem ipsum.
-
+* Fixed various bugs in aggregate methods of Table, TableView and Query for nullable columns.
```
Please spell out which public methods had bugs.

bmunkholm commented on src/realm/column_timestamp.hpp:
```diff
@@ -126,20 +126,21 @@ class TimestampColumn : public ColumnBaseSimple {
 
         Timestamp best = get(0);
-        size_t best_index = 0;
+        size_t best_index = !best.is_null() ? 0 : npos; // npos if null
```
Don't not not avoid changing the negation for readability, as suggested by Daniel :-)

bmunkholm commented:
A few style issues with missing spaces around brackets, so run clang-format.

bmunkholm commented on CHANGELOG.md:
```diff
@@ -2,9 +2,9 @@
-  Non-nullable columns might not have been affected. One of the bugs are described here
-  https://github.com/realm/realm-core/issues/2357
+* Fixed various bugs in aggregate methods of Table, TableView and Query for nullable columns
+  (max, min, avg, sum). Non-nullable columns might not have been affected. One of the bugs are
```
What did they do wrong? Can we be a bit more specific, so we can write a proper release note in SDKs and inform people know how they might have been affected?

rrrlasse commented:
@bmunkholm can you review release notes again now?

bmunkholm commented:
@rrrlasse Is it unclear if non-nullable columns was affected?

rrrlasse commented:
@bmunkholm yes

bmunkholm commented:
👍
